packages/gatsby/src/schema/resolvers.ts

```diff
@@ -75,34 +54,56 @@ export function findOne<TSource, TArgs>(
 
 type PaginatedArgs<TArgs> = TArgs & { skip?: number; limit?: number }
 
-export function findManyPaginated<TSource, TArgs, TNodeType>(
+export function findManyPaginated<TSource, TArgs>(
   typeName: string
 ): GatsbyResolver<TSource, PaginatedArgs<TArgs>> {
   return async function findManyPaginatedResolver(
-    source,
+    _source,
     args,
     context,
     info
-  ): Promise<IGatsbyConnection<TNodeType>> {
+  ): Promise<IGatsbyConnection<IGatsbyNode>> {
     // Peek into selection set and pass on the `field` arg of `group` and
     // `distinct` which might need to be resolved.
     const group = getProjectedField(info, `group`)
     const distinct = getProjectedField(info, `distinct`)
     const max = getProjectedField(info, `max`)
+    const min = getProjectedField(info, `min`)
+    const sum = getProjectedField(info, `sum`)
+
+    // Apply paddings for pagination
+    // (for previous/next node and also to detect if there is a previous/next page)
+    const skip = typeof args.skip === `number` ? Math.max(0, args.skip - 1) : 0
+    const limit = typeof args.limit === `number` ? args.limit + 1 : undefined
+
     const extendedArgs = {
       ...args,
       group: group || [],
       distinct: distinct || [],
       max: max || [],
+      min: min || [],
+      sum: sum || [],
+      skip,
+      limit,
     }
-
-    const result = await findMany<TSource, PaginatedArgs<TArgs>>(typeName)(
-      source,
-      extendedArgs,
-      context,
-      info
+    if (context.stats) {
+      context.stats.totalRunQuery++
+      context.stats.totalPluralRunQuery++
+    }
+    const result = await context.nodeModel.findAll(
+      {
+        query: extendedArgs,
+        type: info.schema.getType(typeName),
+        stats: context.stats,
+        tracer: context.tracer,
+      },
+      { path: context.path, connectionType: typeName }
     )
-    return paginate(result, { skip: args.skip, limit: args.limit })
+    return paginate(result, {
+      resultOffset: skip,
+      skip: args.skip,
+      limit: args.limit,
+    })
```

Comment:
Now that results are already sliced during `findAll` call, pagination can't just do `result.slice(skip, offset)` because it has no idea where `skip` points now. So we pass additional information on the current result offset.

With this information, we can calculate the actual slice start point: `skip - resultOffset`